File: 103. Binary Tree Zigzag Level Order Traversal/solution.js
/** Definition for a binary tree node. */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    const queue = [];
    root && queue.push(root);

    const levels = [];
    let reverse = false;

    while (queue.length > 0) {
        let currentLength = queue.length;
        let currentLevel = [];

        for (let i = 0; i < currentLength; i++) {
            let node = queue.shift();
            if (reverse) {
                currentLevel.unshift(node.val);
            } else {
                currentLevel.push(node.val);
            }

            node.left && queue.push(node.left);
            node.right && queue.push(node.right);
        }
        
        levels.push(currentLevel);
        reverse = !reverse;
    }

    return levels;
};

File: 103. Binary Tree Zigzag Level Order Traversal/solution.py
from typing import *
from collections import deque

class TreeNode:
    """Definition for a binary tree node."""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        q = deque()
        if root:
            q.append(root)
        
        levels = []
        reverse = False
        while q:
            current_length = len(q)
            current_level = []

            for _ in range(current_length):
                node = q.popleft()
                if reverse:
                    current_level.insert(0, node.val)
                else:
                    current_level.append(node.val)
                
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

            levels.append(current_level)
            reverse = not reverse

        return levels